### PR TITLE
WIP: end of twitch

### PR DESCRIPTION
**Pull request by campoy**
why is computeGradient not returning the right values?

### CODE REVIEW

File: 39-linreg-gonum/linreg.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+func computeCost(x, y []float64, m, c float64) float64 {
+	floats.Scale(m, x)    // m * x
+	floats.AddConst(c, x) // m * x + c
+	floats.Sub(y, x)      // y - (m*x + c)
+	return floats.Dot(y, y) / float64(len(x))
+}
+
+func computeGradient(x, y []float64, m, c float64) (dm, dc float64) {
+	floats.Scale(m, x)    // m * x
```
Comment (nebloc) [inline 39-linreg-gonum/linreg.go]: ALL x values after scale are 0

Comment (campoy) [reply 39-linreg-gonum/linreg.go]: if m is 0 yeah, but not otherwise

Comment (campoy) [reply 39-linreg-gonum/linreg.go]: FFS, the problem is that I'm actually changing the value of the points by using `Scale`!

Comment (nebloc) [reply 39-linreg-gonum/linreg.go]: 🙂

Comment (campoy) [reply 39-linreg-gonum/linreg.go]: I guess this is a good friction log for @sbinet and @kunde21

Comment (Kunde21) [reply 39-linreg-gonum/linreg.go]: Or, combine the two:
```
calc = make([]float64, len(y)
...
floats.AddScaledTo(calc, y, -m, x)  // calc = y + (-m * x)
floats.AddConst(-c, calc) // calc = (y + (-m*x) ) .- c

dm = -2 / float64(len(x)) * floats.Dot(x, calc)
dc = -2 / float64(len(x)) * floats.Sum( calc)
return dm, dc
```

Then, you could pass the `calc` slice into both compute functions as a workspace.

**Comment by sbinet** [comment]
running pprof (on the 512000 benchmark):
```
$> go tool pprof ./cpu.prof 
File: 39-linreg-gonum.test
Type: cpu
Time: Aug 1, 2018 at 9:05am (CEST)
Duration: 2.23s, Total samples = 2.12s (94.95%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 2120ms, 100% of 2120ms total
Dropped 3 nodes (cum <= 10.60ms)
Showing top 10 nodes out of 13
      flat  flat%   sum%        cum   cum%
     610ms 28.77% 28.77%      610ms 28.77%  github.com/campoy/justforfunc/vendor/gonum.org/v1/gonum/floats.Sum
     380ms 17.92% 46.70%      380ms 17.92%  github.com/campoy/justforfunc/vendor/gonum.org/v1/gonum/internal/asm/f64.DotUnitary
     310ms 14.62% 61.32%      310ms 14.62%  runtime.memmove
     290ms 13.68% 75.00%      290ms 13.68%  github.com/campoy/justforfunc/vendor/gonum.org/v1/gonum/floats.AddConst
     290ms 13.68% 88.68%      290ms 13.68%  github.com/campoy/justforfunc/vendor/gonum.org/v1/gonum/internal/asm/f64.AxpyUnitaryTo
     240ms 11.32%   100%      240ms 11.32%  github.com/campoy/justforfunc/vendor/gonum.org/v1/gonum/internal/asm/f64.ScalUnitary
         0     0%   100%     2110ms 99.53%  github.com/campoy/justforfunc/39-linreg-gonum.BenchmarkComputeGradient.func1.1
         0     0%   100%     2110ms 99.53%  github.com/campoy/justforfunc/39-linreg-gonum.computeGradient
         0     0%   100%      380ms 17.92%  github.com/campoy/justforfunc/vendor/gonum.org/v1/gonum/floats.Dot
         0     0%   100%      240ms 11.32%  github.com/campoy/justforfunc/vendor/gonum.org/v1/gonum/floats.Scale
(pprof) list Sum
Total: 2.12s
ROUTINE ======================== github.com/campoy/justforfunc/vendor/gonum.org/v1/gonum/floats.Sum in /home/binet/dev/go/gocode/src/github.com/campoy/justforfunc/vendor/gonum.org/v1/gonum/floats/floats.go
     610ms      610ms (flat, cum) 28.77% of Total
         .          .    763:}
         .          .    764:
         .          .    765:// Sum returns the sum of the elements of the slice.
         .          .    766:func Sum(s []float64) float64 {
         .          .    767:	var sum float64
     460ms      460ms    768:	for _, val := range s {
     150ms      150ms    769:		sum += val
         .          .    770:	}
         .          .    771:	return sum
         .          .    772:}
         .          .    773:
         .          .    774:// Within returns the first index i where s[i] <= v < s[i+1]. Within panics if:
```

and the equivalent `loop` one:
```
$> go tool pprof ./cpu.prof 
File: 39-linreg-gonum.test
Type: cpu
Time: Aug 1, 2018 at 9:31am (CEST)
Duration: 1.53s, Total samples = 1.42s (93.02%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 1.42s, 100% of 1.42s total
      flat  flat%   sum%        cum   cum%
     1.42s   100%   100%      1.42s   100%  github.com/campoy/justforfunc/39-linreg-gonum.computeGradientLoop
         0     0%   100%      1.42s   100%  github.com/campoy/justforfunc/39-linreg-gonum.BenchmarkComputeGradient.func1.1
         0     0%   100%      1.42s   100%  testing.(*B).launch
         0     0%   100%      1.42s   100%  testing.(*B).runN
(pprof) list comp
Total: 1.42s
ROUTINE ======================== github.com/campoy/justforfunc/39-linreg-gonum.computeGradientLoop in /home/binet/dev/go/gocode/src/github.com/campoy/justforfunc/39-linreg-gonum/linreg.go
     1.42s      1.42s (flat, cum)   100% of Total
         .          .     37:	dc = 2 * floats.Sum(buf) / n
         .          .     38:	return cost, dm, dc
         .          .     39:}
         .          .     40:
         .          .     41:func computeGradientLoop(buf, x, y []float64, m, c float64) (cost, dm, dc float64) {
      90ms       90ms     42:	for i := range x {
     450ms      450ms     43:		d := y[i] - (x[i]*m + c)
     240ms      240ms     44:		cost += d * d
     400ms      400ms     45:		dm += -x[i] * d
     240ms      240ms     46:		dc += -d
         .          .     47:	}
         .          .     48:	n := float64(len(x))
         .          .     49:	cost = cost / n
         .          .     50:	dm = 2 * dm / n
         .          .     51:	dc = 2 * dc / n

```

I tried to replace the call to `floats.Sum` with `CumSum` (which has an ASM version) to no avail.
not sure we can amortize the cost of going through the slices multiple times with ASM versions of these operations...

File: 39-linreg-gonum/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Linear Regression with Gradient Descent (ML4G #1)
+
+In a [previous episode](https://www.youtube.com/watch?v=ihP7lQivA6M) we learned how to plot data with gonum and we added a line that approximated those points ... but we totally made that up!
+
+During this episode we'll cover how to evaluate how good a line matches the points and how to find the best line possible by using gradient descent.
```
Comment (btracey) [inline 39-linreg-gonum/README.md]: how well a line matches?

File: 39-linreg-gonum/linreg.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+	n := float64(len(x))
+	cost = floats.Dot(buf, buf) / n
+	dm = 2 * floats.Dot(x, buf) / n
+	dc = 2 * floats.Sum(buf) / n
+	return cost, dm, dc
+}
+
+func computeGradientLoop(buf, x, y []float64, m, c float64) (cost, dm, dc float64) {
+	for i := range x {
```
Comment (btracey) [inline 39-linreg-gonum/linreg.go]: You could use the ` for i, v := range x { ` syntax and replace the `x[i]`. My personal style choice is to use `xs, ys []float64` which then allows `for i, x := range xs`.

**Comment by sbinet** [comment]
interestingly, here is the bench with `stat.LinearRegression`:
```
BenchmarkComputeGradient/512000/loop    	    2000	    686744 ns/op	       8 B/op	       0 allocs/op
BenchmarkComputeGradient/512000/floats  	    1000	   1812519 ns/op	       2 B/op	       0 allocs/op
BenchmarkComputeGradient/512000/LR      	     500	   3543072 ns/op	       0 B/op	       0 allocs/op
```

**Comment by btracey** [comment]
We wouldn't expect `LinearRegression` to be faster, as it directly computes the optimal coefficients to (corrected) numerical precision. I just made a PR to optimize the code ( https://github.com/gonum/gonum/pull/563 ), which should make it about 3x the loop version which sounds about right I think.

**Pull request closed without merging**